File: encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/Dispatcher.java
package com.automattic.encryptedlogging;

import org.greenrobot.eventbus.EventBus;
import com.automattic.encryptedlogging.store.Store;
import org.wordpress.android.util.AppLog;
import org.wordpress.android.util.AppLog.T;

import javax.inject.Singleton;


@Singleton
public class Dispatcher {
    private final EventBus mBus;

    public Dispatcher() {
        mBus = EventBus.builder()
                .logNoSubscriberMessages(true)
                .sendNoSubscriberEvent(true)
                .throwSubscriberException(true)
                .build();
    }

    public void register(final Object object) {
        mBus.register(object);
        if (object instanceof Store) {
            ((Store) object).onRegister();
        }
    }

    public void unregister(final Object object) {
        mBus.unregister(object);
    }

    public void dispatch(com.automattic.encryptedlogging.annotations.action.Action action) {
        AppLog.d(T.API, "Dispatching action: " + action.getType().getClass().getSimpleName()
                + "-" + action.getType().name());
        post(action);
    }

    public void emitChange(final Object changeEvent) {
        mBus.post(changeEvent);
    }

    private void post(final Object event) {
        mBus.post(event);
    }
}

File: encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/Dispatcher.kt
package com.automattic.encryptedlogging

import android.util.Log
import com.automattic.encryptedlogging.store.Store
import org.greenrobot.eventbus.EventBus

internal class Dispatcher {
    private val mBus: EventBus = EventBus.builder()
        .logNoSubscriberMessages(true)
        .sendNoSubscriberEvent(true)
        .throwSubscriberException(true)
        .build()

    fun register(`object`: Any?) {
        mBus.register(`object`)
        if (`object` is Store) {
            `object`.onRegister()
        }
    }

    fun unregister(`object`: Any?) {
        mBus.unregister(`object`)
    }

    fun dispatch() {
        Log.d(TAG, "Dispatching")
        post(TODO())
    }

    fun emitChange(changeEvent: Any?) {
        mBus.post(changeEvent)
    }

    private fun post(event: Any) {
        mBus.post(event)
    }

    companion object {
        private val TAG = Dispatcher::class.java.simpleName
    }
}
